sklearn/utils/tests/test_extmath.py

```diff
@@ -161,6 +162,85 @@ def test_randomized_svd_low_rank_all_dtypes(dtype):
     check_randomized_svd_low_rank(dtype)
 
 
+@pytest.mark.parametrize('dtype',
+                         (np.int32, np.int64, np.float32, np.float64))
+def test_randomized_eigsh(dtype):
+    """Test that `_randomized_eigsh` returns the appropriate components"""
+
+    X = np.diag(np.array([1., -2., 0., 3.], dtype=dtype))
```

Comment:
To make the test more interesting, could you please multiply `X` by a random rotation matrix (with a fixed seed), for instance using a `np.linalg.qr` decomposition:

```suggestion
rng = np.random.RandomState(42)

X = np.diag(np.array([1., -2., 0., 3.], dtype=dtype))
# random rotation that should preserve the eigenvalues of X:
X @= np.linalg.qr(rng.normal(size=X.shape))[0]
```

Reply:
Indeed the determinant can be negative (-1) at random. We need a positive orthogonal matrix. Maybe the following would work:

```python
Q = np.linalg.qr(rng.normal(size=X.shape))[0]
if np.linalg.det(Q) < 0:
    Q[0, :] *= -1
X = X @ Q
```

Reply:
It seems that  `Q @ X @ Q.T` does the trick already